### ISSUE TITLE
need some hints as to what short variables mean ?

### ISSUE THREAD

**Issue by teto** [pcapriotti/optparse-applicative#440]
Hi,

I am trying to convert between optparse-applicative and haskeline autocompletion to get autocompletion in haskeline (Ideally I would like to have something similar to https://github.com/python-cmd2/cmd2 but it's a long way)

I find the doc a bit scarce and I have trouble guessing what the short variable names are. I've started haddock documenting parts of optparse code (hopefully there is interest and I can merge those back) but I need some help to understand what "ws" mean in ?
```
haskelineCompletionQuery :: ParserInfo a -> ParserPrefs -> Richness -> [String] -> Int -> String -> IO [String]
haskelineCompletionQuery pinfo pprefs richness ws i _ = case runCompletion compl pprefs of
```
same for `ns` in `filter_names ns` ?

**Comment by HuwCampbell** [comment]
The argument `ws` is the "words" which have been passed in on the command line, as split by the shell. If I recall, we don't actually need `i`, the index of the word which the user pressed tab on, because the completion scripts cut the completion up to the caret position anyway.

In filter_names, `ns` is the names of the sub-commands which the subparser has. The function `filter_names` makes sure that the user could want the completion for a particular command by seeing if the current text at the position of the token is a prefix of the sub command name.

**Comment by teto** [comment]
thanks ! I thought about "words" after postiung this but still was not sure. 
I am now adapting the bashCompletionQuery for haskeline and there is the drop on this line that confuses me:
https://github.com/pcapriotti/optparse-applicative/blob/4fd2325533e74b52cb5c0c25dab619857b6e5ef4/src/Options/Applicative/BashCompletion.hs#L76 , do we drop the first argument because we assume it's the name of the program ? 

Also with the following parser (i.e., a subparser):
```
mainParser :: Parser CommandArgs
mainParser = subparser (
    commandGroup "Generic"
    <> command "help" helpParser
    <> command "quit" quit
    <> commandGroup "Loader commands"
    <> command "load-csv" CL.piLoadCsv
    )
    where
      helpParser = info (pure ArgsHelp) (progDesc "Display help")
```
do you know if bashCompletionQuery would generate completion for all commands ? in my tests, it seems to autocomplete only the first argument but not the subsequent ones because `runParserInfo` returns `Nothing` (while I would expect it to retreive at least one more Parser).

**Comment by HuwCampbell** [comment]
I'm not sure I follow sorry, but yes, the completion system sensibly completes within subparsers as well.

**Comment by teto** [comment]
I've changed optparse-applicative's examples/Cabal.hs to an executable to test the zsh completion.
It seemed to work. Back to my code, I noticed that when typing the full command name, completion would work, but it would fail when the command was half-typed.  I tracked it down to https://github.com/pcapriotti/optparse-applicative/blob/4fd2325533e74b52cb5c0c25dab619857b6e5ef4/src/Options/Applicative/Common.hs#L210
and the following seems to do the trick:
```
diff --git a/src/Options/Applicative/Common.hs b/src/Options/Applicative/Common.hs
index c40f3c4..4edbd05 100644
--- a/src/Options/Applicative/Common.hs
+++ b/src/Options/Applicative/Common.hs
@@ -208,7 +208,9 @@ runParser policy isCmdStart p args = case args of
   (arg : argt) -> do
     (mp', args') <- do_step arg argt
     case mp' of
-      Nothing -> hoistMaybe result <|> parseError arg p
+      Nothing -> 
+        trace "ERROR" hoistMaybe result <|> exitP isCmdStart policy p Nothing
       Just p' -> runParser (newPolicy arg) CmdCont p' args'
   where
     result =
     ```

In terms of API, an idea I like from haskeline   would be to return a completion  datatype(https://hackage.haskell.org/package/haskeline-0.8.2/docs/System-Console-Haskeline-Completion.html#t:Completion)  in
`newtype Completer = Completer   { runCompleter :: String -> IO [String] }` => `newtype Completer = Completer   { runCompleter :: String -> IO [Completion] }` . 

optparse-applicative could return the completion and its description and toggle what to display depending on "Richness" ?

**Comment by teto** [comment]
another thing that is annoying is that let's say I have my positional argument  with a `<> completeWith ["toto", "tata"]` modifier, there is no validation of the resulting value during runCompletion. Let's say my previous command `load-pcap` expects as argument "toto" or "tata"  I wouldl ike to trigger the completion if the value isn't one of those but right now optparse accepts any value.

**Comment by HuwCampbell** [comment]
Yeah... using a type which included an optional richness would be a lot better that just a string which gets a sentinal tab in it. We would allow users to also supply rich completions where they're supported... don't know I didn't do that the first time round.

Obviously until a better API for shells is developed we would have to smash in the tab in eventually, but we would push it to the boundary.

**Comment by teto** [comment]
so I found the issue with eitherReader, it throws an ErrorMsg (which doesn't contian the parser) so runCompletion returns a Nothing instead of the parser with the error. I've hacked the searchArg function so that upon any error it gets rewritten with a ParseError that actually contains the parser:
```
      ArgReader rdr -> let
          tmp = runExcept $ runReaderT (unReadM $ crReader rdr) arg 
        in
        -- rewrite the error so that completion can access the parser
        fmap pure . lift . lift $ either (const $ exitP CmdStart AllPositionals parser Nothing) return tmp
```
This way I can access the parser that failed during autocompletion, and display the available completions. This hack may break other stuff but it solves the last (big) issue I had with live completion. I am running after a deadline but once I get more time, I can submit PRs if there is interest. Out of the box completion with haskeline would be a neat feature IMO (or at least making it easier).
